GooglePlayInstant/Editor/ApkSigner.cs

```diff
@@ -0,0 +1,184 @@
+{
+    /// <summary>
+    /// Provides methods that call the Android SDK build tool "apksigner" to verify whether an APK complies with
+    /// <see href="https://source.android.com/security/apksigning/v2">APK Signature Scheme V2</see> and to re-sign
+    /// the APK if not. Instant apps require Signature Scheme V2 starting with Android O, however some Unity versions
+    /// do not produce compliant APKs. Without this "adb install --ephemeral" on an Android O device will fail with
+    /// "INSTALL_PARSE_FAILED_NO_CERTIFICATES: No APK Signature Scheme v2 signature in ephemeral package".
+    /// </summary>
+    public static class ApkSigner
+    {
+        private const string AndroidDebugKeystore = ".android/debug.keystore";
+
+        /// <summary>
+        /// Returns true if apksigner is available to call, false otherwise.
+        /// </summary>
+        public static bool IsAvailable()
+        {
+            return GetApkSignerJarPath() != null;
+        }
+
+        /// <summary>
+        /// Synchronously calls the apksigner tool to verify whether the specified APK uses APK Signature Scheme V2.
+        /// </summary>
+        /// <returns>true if the specified APK uses APK Signature Scheme V2, false otherwise</returns>
+        public static bool Verify(string apkPath)
+        {
+            var arguments = string.Format("-jar {0} verify {1}", GetApkSignerJarPath(), apkPath);
+            var result = CommandLine.Run(JavaUtilities.JavaBinaryPath, arguments);
+            if (result.exitCode == 0)
+            {
+                return true;
+            }
+
+            Debug.LogWarningFormat("\"java {0}\" failed with exit code {1}", arguments, result.exitCode);
+            return false;
+        }
+
+        /// <summary>
+        /// Synchronously calls the apksigner tool to sign the specified APK using APK Signature Scheme V2.
+        /// </summary>
+        /// <returns>true if the specified APK was successfully signed, false otherwise</returns>
+        public static bool Sign(string apkPath)
+        {
+            string keystoreName;
+            string keystorePass;
+            string keyaliasName;
+            string keyaliasPass;
+            if (string.IsNullOrEmpty(PlayerSettings.Android.keystoreName))
+            {
+                Debug.LogFormat("No keystore specified. Signing using {0}.", AndroidDebugKeystore);
+                var homePath =
+                    Application.platform == RuntimePlatform.WindowsEditor
+                        ? Environment.ExpandEnvironmentVariables("%HOMEDRIVE%%HOMEPATH%")
+                        : Environment.GetEnvironmentVariable("HOME");
+                if (string.IsNullOrEmpty(homePath))
+                {
+                    Debug.LogErrorFormat("Failed to locate home directory that contains {0}.", AndroidDebugKeystore);
+                    return false;
+                }
+
+                keystoreName = Path.Combine(homePath, AndroidDebugKeystore);
```

Comment:
should we check to see if this actually exists?

Reply:
It doesn't fail fast. It prompts for the password before failing with a FileNotFoundException. It's pretty noisy, so I will add the File.Exists() check as you suggest.